### ISSUE TITLE
Refactor CSV import code to be more efficient

### ISSUE THREAD

**Issue by NealHumphrey** [codefordc/housing-insights#54]
------------------------
update

This has been partially completed and is in the overhaul-csv-loading branch. 

The remaining tasks that need to be completed are 
#83 
#86

This ticket should be closed, but keeping in the 'on hold' column so it is available for reference if needed. 

----------------------------------------------------
## 4 parallel tasks that we can do now:
- [x] `(salomone)` Make JSON data: We will store our list of expected fields and their data types into a file called meta.json. Use Pandas to take a new .csv file, guess the data types and produce the meta.json. Use the `make_draft_json.py` file as a starting point. 

- [x] `(Brandon)` Figure out how to read a CSV into memory (so that we can edit values such as messy data) and then write it to SQL efficiently. Use `demo_load_to_sql_without_pandas.py` as a starting point. 

- [ ] `(unclaimed)` Everything else for our `load_data.py` file, which handles downloading files if needed, comparing manifest.csv to the manifest table in SQL to decide whether to load a table or not, handling errors. Temporarily just use dataframe = pandas.read_csv and dataframe.to_sql() in the csv_to_sql() function until #2 is ready. Use my list of functions and the rough outline in the main() function as a guide, but there are probably additional things I didn't think of. 

- [x] `(Neal)` Get data upload tools to S3 configured. 

Stuff to do after:
- [ ] run make_draft_json.py on all our data
- [ ] set up data cleaning phase in between read and write (see note in the `csv_to_sql()` function)

To get working:
- Pull the branch `overhaul-csv-loading` into your repo
- Make a new branch off of that branch for your portion
- When you're ready to share your work, make a pull request on Github into the `overhaul-csv-loading` branch (*not* the dev branch). We'll consolidate all our work there and make sure it's working before merging into dev. 
(let me know if you need help with any of this!)

## *What you'll see in the /scripts folder:*
**/`__root__`**
- database_management.py: contains tools to connect to our database. Moved to root because this can be also be used later in other places (e.g. analysis scripts). I created this for housing_risk but I have not yet tested this in our project, so it might need minor changes. 
-secrets.example.json: check out the addition of the local_database vs. remote_database, and make this edit to your secrets.json copy (or ask me for a copy if you don't have one yet).

**/data**
- I moved the small_data, big_data, and output_data folders into here to clean up our root

**/ingestion:**
- make_draft_json.py: for item #1
- demo_load_to....: for item #2
- load_data.py: for item #3
- manifest.csv: I edited the columns available here to match what was used in the housing_risk example code. I deleted all the rows except for 3 - one small table that is immediately available in the repo, plus two large tables that are both a) able to be downloaded from S3 (to test download code) and b) need to go to the same table in SQL (to test append to sql table functionality)
- meta.json: my proposed structure for the JSON file that stores field names and data types.
- example_....: these are the old load_data.py scripts and files, which you can look at for reference on how things are previously done. We'll delete these once we get everything working.

If you're feeling ambitious:
Unit tests!!!
Since this new approach should be production-worthy, I want to make sure we get at least a decent amount of unit tests written. If you're into test driven development, write these tests as you go, otherwise I'll want to get them written before we merge to dev. If you're not familiar with unit tests, don't worry about it yet. 

**Resources**:
* High performance with named tuples and custom classes, but only discusses the read from csv part of the equation. Slightly clunky for data type parsing. Typically we won't have larger-than-memory data sets so the SQL inserts are our bigger problem. [District Data Labs CSV article](https://districtdatalabs.silvrback.com/simple-csv-data-wrangling-with-python)
* [Odo](http://odo.pydata.org/en/latest/overview.html), another library that claims to be good at parsing dates. 
* BULK INSERT command in SQL. Use this with the load into Pandas, fix data problems (cleaning), and then could process comma delimited into pipe delimited flat file to deal with some data format issues and load from there? [StackOverflow Discussion](http://stackoverflow.com/questions/29638136/how-to-speed-up-with-bulk-insert-to-ms-server-from-python-with-pyodbc-from-csv/29649340#29649340). Pay attention to the dtype parameter in the pandas.read_csv method to pass it manual overrides of data types instead of inferring when reading the file.

**Comment by salomoneb** [comment]
Neal, 

In `make_draft_json.py`, you [note](https://github.com/codefordc/housing-insights/blob/overhaul-csv-loading/scripts/ingestion/make_draft_json.py#L54): 

> #use command line arguments to pass (relative) filepath and table_name.
> #we should use positional arguments in the command line

Just to confirm, the code will only ingest and convert the files that are given as command line arguments, correct?

If so, is this how you picture the command to run the program?
`python make_draft_json.py ../example/relative/filepath example_table`

Should the program be able to parse multiple arguments, ie.
`python make_draft_json.py ../example/relative/filepath1 example_table1, ../example/relative/filepath2 example_table2`

[argparse](https://docs.python.org/3/library/argparse.html) seems like it might be good for this. Alternatively, I wonder if we could use [fileinput](https://docs.python.org/3/library/fileinput.html) to read the information from the manifest file. This might be cleaner than specifying the data files on the command line (but could also maybe get confusing)? What do you think?

**Comment by NealHumphrey** [comment]
@salomoneb - I think that it is OK to use just one file path for the command line arguments. I think normally we will be processing files one-by-one (since we have to manually review the output anyways...). 

If we do have a need to run multiple files frequently, we can import this code into another file and have it call the make_draft_json() function directly - the `if '__name__' == '__main__' code is only run when this file itself is run from the command line. 

I don't think we should add the files to the existing manifest.csv until they have been "validated", i.e. the file has been added to the meta.json (by running the make_draft_json routine). 

Argparse looks like it will be handy if/when we start getting complex. For now, I think we can just run with:
`sys.argv[1]` - returns the first item after the make_draft_json.py name
`sys.argv[2]` - returns the second item after the make_draft_json.py name

**Comment by salomoneb** [comment]
Neal, 

https://gist.github.com/salomoneb/53f8c6fb851346f7dd812fa518808859

Here's a test/proof-of-concept version of this script. There are still two things left to do with this: 

1. Error handling and cleaning
2. Expanding the metadata for each field

Regarding the second point, can you explain the difference between `field_name_matches_sql_name`, `source_name`, and `sql_name`? Also, I assume `display_name` and `display_text` would be manually added?

**Comment by NealHumphrey** [comment]
Cool.

Your questions:
`"source_name"` - whatever the field is called in the original CSV file. If the first row of the CSV is a header row, this is the string that matches; if there is no header, it is an integer. 
`sql_name` = if the name in the original CSV is not intuitive, we can optionally call it something else in our database (e.g. ACS rent is 'hd01_vd01' instead of 'median_rent'). The make_json script should generate a default by taking `source_name` and applying functions to make it play nice with SQL, e.g. replace space with underscore, force to lower case, etc. 
`display_name` - yes, added manually - but your script should pre-fill with sql_name / source_name. 
`display_text` - yes, add manually

So basically your script should make source_name, sql_name, and display_name all equal to whatever the column is named in the CSV - having 3 versions is to give us flexibility.

**Comment by NealHumphrey** [comment]
@blw9u2012 here are the 5 tables for the Preservation Catalog, which are a good starting point for testing real data: https://s3.amazonaws.com/nhumphrey-misc/PresCat_Export_20160401.zip

**Comment by salomoneb** [comment]
Photos of the workflow that we discussed at our meeting last night. 
![img_3290](https://cloud.githubusercontent.com/assets/16308025/23223604/76fe3944-f8f9-11e6-86fb-2bd351785c3d.JPG)
![img_3292](https://cloud.githubusercontent.com/assets/16308025/23223615/7c43e778-f8f9-11e6-9287-af1a60ed78d6.JPG)
![img_3293](https://cloud.githubusercontent.com/assets/16308025/23223618/7de38e6c-f8f9-11e6-85cf-b6c4ae2ac8e1.JPG)
![img_3295](https://cloud.githubusercontent.com/assets/16308025/23223625/8510c740-f8f9-11e6-8530-96b532fc4ea5.JPG)
